client.go

```diff
@@ -0,0 +1,89 @@
+package main
+
+import (
+	"log"
+	"time"
+
+	"github.com/gorilla/websocket"
+)
+
+const (
+	pongPeriod = 60 * time.Second
+	pingPeriod = 30 * time.Second
+)
+
+type Client struct {
+	conn *websocket.Conn
+	out  chan OutFrame
+}
+
+func NewClient(conn *websocket.Conn) *Client {
+	return &Client{
+		conn: conn,
+		out:  make(chan OutFrame),
+	}
+}
+
+func (c *Client) tx() {
+	ticker := time.NewTicker(pingPeriod)
+
+	defer func() {
+		ticker.Stop()
+		c.conn.Close()
+	}()
+
+	for {
+		select {
+		case frame, ok := <-c.out: // Regular frames
+			if !ok {
+				c.conn.WriteMessage(websocket.CloseMessage, []byte{})
+				log.Println("CloseMessage")
+				return
+			}
+
+			// Send the outbound frame as JSON
+			if err := c.conn.WriteJSON(frame); err != nil {
+				log.Println("error: ", err)
+				return
+			}
+
+		case <-ticker.C: // Ping frames
+			if err := c.conn.WriteMessage(websocket.PingMessage, nil); err != nil {
+				return
+			}
+		}
+	}
+}
+
+func (c *Client) rx(gameInChannel chan InFrame) {
+	defer func() {
+		c.conn.Close()
+	}()
+
+	c.conn.SetReadLimit(maxMessageSize)
```

Comment:
Do you need to do this on each receive?

Reply:
I don't think this is on each receive; I think this is invoked once for each websocket connection (and then runs forever in the infinite loop below this), and I think I do need to set this for each connection once.